x/valset-pref/validator_set.go

```diff
@@ -119,72 +128,125 @@ func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, co
 // For ex: userA has staked 10tokens with weight {Val->0.5, ValB->0.3, ValC->0.2}
 // undelegate 6osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
 // our undelegate logic would attempt to undelegate 3osmo from A, 1.8osmo from B, 1.2osmo from C
-// nolint: staticcheck
+// Rounding logic ensures we do not undelegate more than the user has staked with the validator set.
+// NOTE: check readme.md for more verbose description of the algorithm.
 func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string, coin sdk.Coin) error {
-	// get the existingValSet if it exists, if not check existingStakingPosition and return it
 	existingSet, err := k.GetDelegationPreferences(ctx, delegatorAddr)
 	if err != nil {
 		return fmt.Errorf("user %s doesn't have validator set", delegatorAddr)
 	}
 
-	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
-	if err != nil {
-		return err
-	}
-
+	delegator := sdk.MustAccAddressFromBech32(delegatorAddr)
 	// the total amount the user wants to undelegate
 	tokenAmt := sdk.NewDec(coin.Amount.Int64())
 
-	err = k.CheckUndelegateTotalAmount(tokenAmt, existingSet.Preferences)
-	if err != nil {
-		return err
-	}
-
-	// totalDelAmt is the amount that keeps running track of the amount of tokens undelegated
-	totalUnDelAmt := sdk.NewInt(0)
-	amountToUnDelegate := sdk.NewInt(0)
-
-	for i, val := range existingSet.Preferences {
+	// Step 1-2
+	var valSetRatio []ValRatio
```

Comment:
Why did we remove the helper I had before? It improves gas costs by removing repeated Get's with GetDelegationPreferences, and makes this main function cleaner?

Reply:
i did this for couple of reason 
1. the helper func was useful except i wanted current delegations to be represented as valset {valAddr, weight} and the helper func did the opposite. (represented valset as current delegation)
2. I checked the flow and current func i'm using has less Get calls to staking keeper than the helper function.  (helper = getValset -> if exist -> getDelegation -> else -> GetDelegatorDelegation, nonHelper = getValSet -> if exist return -> else -> GetDelegatorDelegations)

Reply:
Ah interesting, didn't notice the delegations call being reused my bad

Reply:
Err, the delegations aren't getting reused rightnow if you don't have a validator set. This is causing more gas usage.

Reply:
if you dont have a validator set but have existing delegations we create a validator set for you with existing delegations 
ref: https://github.com/osmosis-labs/osmosis/pull/5967/files#diff-eca64719739dd13c273495f6d28e8da96083fdf0c629da0eef1a22206839ed24R84